src/FwdState.cc

```diff
@@ -1308,9 +1308,11 @@ getOutgoingAddress(HttpRequest * request, Comm::ConnectionPointer conn)
             else
 #endif
                 conn->local = request->client_addr;
+
             // some flags need setting on the socket to use this address
             conn->flags |= COMM_DOBIND;
             conn->flags |= COMM_TRANSPARENT;
+            conn->local.port(0); // allow random outgoing port to prevent address clashes
```

Comment:
Please move this line up, right after the `local` assignment, without an empty line in-between: We are _fixing_ that assignment, and the `local` value remains wrong until we reset the port. We should keep this fix as close to the assignment it is fixing as possible...

---

This assignment-fixing logic is repeated in many places, including ClientRequestContext::clientAccessCheck(), ftpOpenListenSocket(), Eui::Eui48::lookup(), IcmpSquid::Recv(), Ftp::Server::listenForDataConnection(), and Ftp::Server::createDataConnection(). We could add and use a dedicated Ip::Address method:

```
/// makes us the same as addr except for a possibly different port
void reset(const Address &addr, const int port);
```

but that method does not help avoid similar bugs because buggy code can still mindlessly copy the whole Ip::Address using other public methods and operators. Thus, I doubt we should add this or a similar method. On the other hand, using reset() would make the code a little cleaner when there are two if/else address assignments (as there are in our case). Your call.